### PR TITLE
Fix Allow header according to RFC 7231

### LINKED ISSUE #2397: 405 Method Not Allowed is missing the Allow header.
## Description

https://tools.ietf.org/html/rfc7231#section-6.5.5 says

>The origin server MUST generate an Allow header field in a 405 response containing a list of the target resource's currently supported methods.

gin does not do this with `HandleMethodNotAllowed` set to `true`. Also it is arguable that even with this setting enabled, all HTTP/1.0 requests should still be given a 404 error as HTTP 1.0 does not define 405.

## How to reproduce

Server:
```
package main

import (
	"net/http"

	"github.com/gin-gonic/gin"
)

func main() {
	router := gin.Default()
	router.GET("hi", func(c *gin.Context) {
		name := c.Param("name")
		c.String(http.StatusOK, "Hello %s", name)
	})
	router.HandleMethodNotAllowed = true
	router.Run()
}
```

Test:
```
$ curl -D- localhost:8080/hi -XPOST
```

## Expectations

```
HTTP/1.0 405 Method Not Allowed
Allow: GET
```

## Actual result

```
HTTP/1.0 405 Method Not Allowed
Content-Type: text/plain
Date: Mon, 25 May 2020 21:53:14 GMT
Content-Length: 22

405 method not allowed
```

## Environment

- go version: go version go1.14.3 darwin/amd64
- gin version (or commit ref): v1.6.3
- operating system: Mac

### PR DESCRIPTION
**Pull request by unbyte**
> 6.5.5.  405 Method Not Allowed
>
>   The 405 (Method Not Allowed) status code indicates that the method
>   received in the request-line is known by the origin server but not
>   supported by the target resource.  The origin server MUST generate an
>   Allow header field in a 405 response containing a list of the target
>   resource's currently supported methods.

close #2397

### CODE REVIEW

File: gin.go [hunk 1]
```diff
@@ -434,16 +435,23 @@ func (engine *Engine) handleHTTPRequest(c *Context) {
...
 			if value := tree.root.getValue(rPath, nil, unescape); value.handlers != nil {
-				c.handlers = engine.allNoMethod
-				serveError(c, http.StatusMethodNotAllowed, default405Body)
-				return
+				allowed = append(allowed, tree.method)
 			}
 		}
+		if len(allowed) > 0 {
+			c.handlers = engine.allNoMethod
+			c.writermem.Header()["Allow"] = []string{strings.Join(allowed, ", ")}
```
Comment (rickb777) [inline gin.go:451]: should this be
```go
c.writermem.Header().Set("Allow", strings.Join(allowed, ", "))
```
?

File: gin.go [hunk 2]
```diff
@@ -434,16 +435,23 @@ func (engine *Engine) handleHTTPRequest(c *Context) {
...
+			serveError(c, http.StatusMethodNotAllowed, default405Body)
+			return
+		}
 	}
 	c.handlers = engine.allNoRoute
```
Comment (rickb777) [inline gin.go:456]: I'd like a blank line after 455 - a change of subject matter.

**Review by rickb777 — COMMENTED** [review]
Looks good to me (with some suggested edits).

File: gin.go (hunk 1 above)
Comment (thinkerou) [reply gin.go:451]: Yes, should use `Set`

**Comment by antonjah** [comment]
@unbyte @rickb777 Any news on this?

**Comment by appleboy** [comment]
@antonjah conflicts

**Comment by Crocmagnon** [comment]
I opened a new PR using @unbyte's commits but fixing conflicts, hoping this can get merged & released soon :)

**Comment by appleboy** [comment]
closed PR and move to https://github.com/gin-gonic/gin/pull/3759

**Pull request closed without merging**